### PR TITLE
Starred courses filter

### PR DESCRIPTION
**Pull request by azito122**
This is a preliminary implementation. I'm sure there is a lot of fine tuning to be done.

I anticipate particular difficulty in the following areas:

**1. The star link**

....which is added to the rubric's 'label'. This is a link that shows up whenever the user is on a course page, and clicking it stars or unstars the course. 

Adding that link to the label necessitated that I alter `rubric.mustache` to accept html in the label tag. This seems less than ideal. I'm not 100% sure what the ideal solution would be for including the star link, but perhaps the rendering/templating system could help. Maybe give filters the ability to add an override template file for rubric and/or list_item?

**2. Live reloading of courses**

Right now, after you click the star link, it will update it's appearance to match the new course state (filled for starred, outline for unstarred). However, the course itself will not show up in the rubric until the page is reloaded. It seems highly desirable, if not essential (especially since we're doing an ajax call already anyway), to reload the displayed courses for the filter when the user stars or unstars.

If we integrate with the templating system as suggested above, then that may help. It is possible to reload templates from JavaScript, so potentially we could re-render the rubric template on that ajax call.

**3. Accessibility**

How can we make this filter more accessible? It's a link, so we can tab to it at least, but I'm not sure that having it be in the label at all is optimal. Also there may be aria tags that should be added to the link and/or icon element.

**4. External functions**

In order to implement the ajax call in a standards compliant way, it was necessary to add `db/services.php` and `externallib.php`, basically creating a whole web service for that one request. It's a little unfortunate that that much scaffolding is required, but on the other hand, it is not totally inconceivable that some external system might want the ability to star/unstar a course, so I don't think it's a total waste. Besides, it's the Moodle way to do things. That said, I have no qualms about rewriting the ajax call and the backend to be less verbose, if we prefer brevity/neatness over the Moodle standards in this case.

I am opening this pull request despite the above issues, in order to start the conversation.

A screenshot of the block in an unstarred course:
starred-courses-filter-0

A screenshot of the block in that same course, after the user has starred it:
starred-courses-filter-1

### CODE REVIEW

**Comment by kwiliarty** [comment]
I haven't looked at the code yet; just thinking about the UX for a start. If I understand correctly, the way the starring mechanism is working now is:

1. The star is on the rubric title
1. Whether it's filled or not depends on whether the current course has been starred?

It's the second issue I'm unclear about, but if I've understood it correctly, then I also foresee this problem: We have no guarantee that an installation is putting the FCL block in every course. An institution might use it only on the site home or "my" page, for instance. In that situation it would be impossible to star any courses. When you're in the course there's no block.

What I had imagined from your earlier descriptions is a scenario where the stars appear next to every course in the block, but only if that block is using a starred filter. Each star is filled or empty depending on whether the given course has been starred. All the courses under the starred rubric would be filled in, but you could click on one of those to remove it. Under other rubrics you could have any combination of filled or empty stars. Clicking on any star would toggle the state of the course for that user, and we can use Ajax to update it. (Starred courses would continue to appear under their original rubrics, too). 

How does that sound?

Let's go the Moodle-y way with the AJAX. 

This is also making me think that we may want another block that could

**Comment by azito122** [comment]
Yes, the star is filled on the course page of a course that is starred, and outlined on non-starred course pages. On non-course pages, it does not appear.

I'm afraid that having a star next to every course would be cluttered. Also, if the user's FCL block isn't enormously long, it could be redundant to have starred courses showing up twice. However, you're absolutely right that if the block is not displayed on course pages, then we need a way for the user to star courses.

I suggest that we make this an option for the filter -- eg, do you want the star link displayed in the rubric title, or by each course in the block? Because I think there is a valid use case for both.

I'd also like to float the idea of having star links that are displayed by course names be hidden by default, and shown when the user hovers over the course name. I think this would prevent visual clutter.

Finally, I will note that displaying a star link by every course name will be a bit tricky to code neatly, especially if we're trying to keep the filters separate. However, I wonder if allowing the definition of overridden templates would help with this -- eg, starred courses could define an overridden `list_item` template that includes the star link. That way there would still be clear separation.

P.S. I will also float the idea of a "star menu" -- eg, the star link in the rubric title opens a select menu, and you can select courses to star, or something. I don't really think that's the right direction, but I want to put it out there in case it sparks a different idea.

**Comment by kwiliarty** [comment]
It's good to just put ideas out there at this point. 

To you point about visual clutter: I was thinking that we could ditch the usual course icon if we are using stars. 

To the question of allowing filters to override templates: That's only necessary if we're trying to treat the starred filter just like any other. I would be okay with saying that it's special. There are other special filters already: "Other courses" and "generic." They are filters, and we get to reuse filtering code for them, but they behave in ways that have global implications. I think that the starred filter may be in a similar situation. For instance, it doesn't really make sense to have more than one of them. If we decide that the starred filter is special, then we can allow ourselves to do things for it without feeling compelled to offer other filters something similar. Food for thought.

Courses showing up more than once: It is mostly how the block already works. There are exceptions -- this is one of the ways the "Other courses" filter is different. The idea is that a "Fall 2018" filter should show all the user's "Fall 2018" courses. It would be misleading to remove a course just because it happens to be starred. This does raise a fun edge case, though... What happens if you star a course in the "Other courses" rubric? In that case, I would agree that it should then be removed. Because of the way "Other courses" work, that should happen naturally. Other courses always comes last and it only shows courses that have not already been listed elsewhere.

This is good discussion. It might make a good Hack-Doc lunch session. Or we should have a breakout for it where dockers could weigh in, too.

**Comment by azito122** [comment]
So thinking about this some more, and specifically thinking about how Starred Courses fits in with other filters in the FCL -- my conclusion is that, in fact, it doesn't. As you pointed out, Starred Courses is special -- there should only be one instance, it may need to alter block-wide markup, and it does a lot more than any other filter. As far as I'm aware, existing filters do exactly that, they _filter_ courses based on **_existing_** course data. Starred Courses both adds course data and provides a special UI interface for interacting with that data.

Therefore I propose two possible paths forward, which provide advantages in terms of both functional and code-level organization:

1. Have a Starred Courses filter, but all it will do is find courses marked as Starred and display them. The actual work of allowing users to star courses, keeping track of courses that are starred, etc., will be split out into a local plugin, e.g. `local_starred_courses`. That plugin will provide a way to star/unstar courses, perhaps a link in an existing navigation menu. That plugin will provide functions like `get_starred_courses()` which will allow FCL to pull in starred courses and display them easily, without corrupting its own code with a bunch of functionality that doesn't fit with the existing plugin. This also provides the opportunity for future plugins to interact with starred courses more easily, if someone else wants to display them in a different way. Obviously the disadvantage is that this adds a dependency for an FCL filter.

2. Instead of a filter, have a setting, e.g. "Display starred courses?" The rest of the functionality flows from there. Instead of being a filter, Starred Courses code will live somewhere else -- or even just as a class file, it doesn't really matter. The point is that it doesn't extend the `filter` class, and that it isn't bound by the `filter` interface. It will be a truly separate entity. This is a less extreme option than splitting the course starring functionality out into a separate plugin, but still provides some of the benefits of not trying to shoehorn Starred Courses into the existing filter interface.

I have a preference for solution 1, but I would prefer either of these, or some combination or evolution, over continuing to try and fit Starred Courses into the filter interface.

**Comment by kwiliarty** [comment]
Since you are doing the development it is of course up to you, but allow me to make the case for continuing on the current path. Let me start by suggesting that we do not have to focus solely on whether starred courses is a filter or not; it's a false dichotomy. I would say that starred courses does some filtering, but is not _only_ a filter. It also represents a welcome overall enhancement to the FCL block. Now here are a bunch of specific observations:

* Starred courses filters a user's courses and displays them in a list of some kind. So much of that basic structure is already there in the FCL that it would seem a shame to me not to take advantage of that. 
* From the UX perspective I think it would be much nicer to have a single interface that shows favorite courses and courses by semester (or any other filtering mechanism) all in one place.
* From a Moodle admin perspective I find it nicer not to have two very similar plugins and not to have to display two blocks. It is also nicer to be able to configure everything in one place.
* From a CLAMP perspective it is great if you and I provide some redundancy for the FCL.
* The FCL is already an approved plugin with a respectable base and a comprehensive testing and documentation infrastructure. There's a lot of work that would not need to be recreated, and I think it provides a good platform for getting this functionality into the hands of the users. 
* You mentioned the idea of a 'starring menu' above. The way I'm thinking of it, that's just what the block would be.
* A boolean option for showing stars lacks some important benefits. If we use a configline then the admin can 
  * label it as desired
  * set a preferred expansion state
  * order it as desired
* I would not worry about the fact that we personally can't envision a use case where it would make sense to list it more than once. I would just allow it. Let admins use their own judgment.
* Any web service that you are thinking of offering and any data that you are thinking of storing can be done just as well from the FCL as from an independent module of some kind. 
* It is no problem to modify our templates as needed, even if we don't intend to make the same kind of accommodation for every filter. 
* None of this is to say that there can't be any other related modules. I can imagine a separate block or activity module, for instance, that would let a user star the current course. 

Summing up: I think that starring would be a great enhancement to the FCL and I would be delighted to have it happen here. Are there some issues to resolve? Sure, but I am very confident that they are all solvable. We don't need to be splitting this functionality across modules. I would even venture to say that we owe it to our community to keep it together. That's my two cents.

**Comment by azito122** [comment]
I agree with a lot of what you say. But I think I also failed to communicate my vision clearly enough. Specifically, I want to note that if we were to split off Starred Courses as `local_starred_courses`, that plugin would NOT be displaying those courses. That functionality would be part of FCL, as the `starred` filter.

> Starred courses filters a user's courses and displays them in a list of some kind. So much of that basic structure is already there in the FCL that it would seem a shame to me not to take advantage of that.

We can still take advantage of that. 

If we use a `starred` filter that reads from `local_starred_courses`, then that will just be a regular filter and can use the whole system -- rubrics, templates, etc. (And note that it won't _need_ any kind of special allowances, because in that case it won't be doing anything other filters don't do -- that functionality would be in `local_starred_courses`). 

If we were to do something more like Solution 1, and make Starred Courses (SC) a separate, non-`filter` class or something, we still COULD use rubrics, templates, etc., but we wouldn't be bound by the conventions in use for other `filter` classes. And, we wouldn't have to worry about making allowances in the filter system for SC, allowances that the other filters don't need and that needlessly complicate the filter system.

> From the UX perspective I think it would be much nicer to have a single interface that shows favorite courses and courses by semester (or any other filtering mechanism) all in one place.

In terms of the course display, it would still all be in one place. In terms of the admin interface, you're right, if we used either Solutions 1 or 2 it would be split in some way -- either between filters and a SC setting, or to some extent between FCL and `local_starred_courses`. However, `local_starred_courses` would not need to have very much of an interface, I don't think.

> From a Moodle admin perspective I find it nicer not to have two very similar plugins and not to have to display two blocks. It is also nicer to be able to configure everything in one place.

Again, I am NOT imagining that SC would appear as a separate block. As to configuring everything in one place, I agree, but I don't think there is much configuring to be done in terms of `local_starred_courses` -- all the display settings would still be in FCL.

> The FCL is already an approved plugin with a respectable base and a comprehensive testing and documentation infrastructure. There's a lot of work that would not need to be recreated, and I think it provides a good platform for getting this functionality into the hands of the users.

I absolutely agree. FCL is an exemplary plugin in many ways. One of those ways is that it is very clear about what it does -- it filters courses by existing data and displays them. As it is an exemplary plugin, I would rather not see its purity of purpose polluted by the SC functionality. That might seem like a nitpick, but to my mind, it is an essential part of maintaining a strong plugin.

As to work being recreated, again, there would be no overlap between the functionality of FCL and `local_starred_courses`.

> A boolean option for showing stars lacks some important benefits. If we use a configline then the admin can
>     label it as desired
>     set a preferred expansion state
>     order it as desired

That's fair. On the other hand, we could add a setting for SC label. As to ordering and collapsibility, I'm not even sure those are things that are essential to SC. If you're not putting it at the top, then what's the point? It's for quick access. If it's collapsed by default, again, what is the point?

However, I can also see that we would want to provide as much flexibility as possible. That's one reason I'd like to split SC into `local_starred_courses` -- then the `starred` filter can just be a regular filter and take advantage of all of the above.

> Any web service that you are thinking of offering and any data that you are thinking of storing can be done just as well from the FCL as from an independent module of some kind.

Well, technically speaking, yes. But if someone decides that they want their own block that displays starred courses, and they want it to operate differently than FCL, then having the SC functionality locked up inside of FCL isn't ideal. I really feel that starring/unstarring courses and storing that user preferences is something that is outside the scope of FCL.

**Comment by kwiliarty** [comment]
Based on our discussions I'm going to close this for now because I believe that you are going to be taking a different approach. I'm happy to entertain any new PR's when you have them.
